core/server/common/src/main/java/alluxio/master/journal/raft/SnapshotReplicationManager.java

```diff
@@ -0,0 +1,503 @@
+  private static final Logger LOG = LoggerFactory.getLogger(SnapshotReplicationManager.class);
+  private static final long SNAPSHOT_REQUEST_TIMEOUT_MS =
+      ServerConfiguration.getMs(PropertyKey.MASTER_EMBEDDED_JOURNAL_TRANSPORT_REQUEST_TIMEOUT_MS);
+
+  private final SimpleStateMachineStorage mStorage;
+  private final RaftJournalSystem mJournalSystem;
+  private volatile long mSnapshotRequestTime = 0L;
+  private volatile RaftJournalServiceClient mJournalServiceClient;
+  private volatile SnapshotInfo mDownloadedSnapshot;
+
+  private enum DownloadState {
+    /** No snapshot download is in progress. */
+    IDLE,
+
+    /** Snapshot information is requested from available followers. */
+    REQUEST_INFO,
+
+    /** The latest snapshot data is requested from one of the followers. */
+    REQUEST_DATA,
+
+    /** The latest snapshot is being downloaded from one of the followers. */
+    STREAM_DATA,
+
+    /** A snapshot is downloaded. */
+    DOWNLOADED,
+
+    /** A snapshot is being installed to the journal storage. */
+    INSTALLING,
+  }
+
+  private final AtomicReference<DownloadState> mDownloadState =
+      new AtomicReference<>(DownloadState.IDLE);
+
+  /**
+   * @param journalSystem the raft journal system
+   * @param storage the snapshot storage
+   */
+  public SnapshotReplicationManager(RaftJournalSystem journalSystem,
+      SimpleStateMachineStorage storage) {
+    mStorage = storage;
+    mJournalSystem = journalSystem;
+  }
+
+  /**
+   * @param journalSystem the raft journal system
+   * @param storage the snapshot storage
+   */
+  @VisibleForTesting
+  SnapshotReplicationManager(RaftJournalSystem journalSystem,
+      SimpleStateMachineStorage storage, RaftJournalServiceClient client) {
+    this(journalSystem, storage);
+    mJournalServiceClient = client;
+  }
+
+  /**
+   * Downloads and installs a snapshot from the leader.
+   *
+   * @return a future with the term index of the installed snapshot
+   */
+  public CompletableFuture<TermIndex> installSnapshotFromLeader() {
+    if (mJournalSystem.isLeader()) {
+      return RaftJournalUtils.completeExceptionally(
+          new IllegalStateException("Abort snapshot installation after becoming a leader"));
+    }
+    if (!transitionState(DownloadState.IDLE, DownloadState.STREAM_DATA)) {
+      return RaftJournalUtils.completeExceptionally(
+          new IllegalStateException("Illegal state while installing a snapshot"));
+    }
+    try {
+      RaftJournalServiceClient client = getJournalServiceClient();
+      SnapshotDownloader<DownloadSnapshotPRequest, DownloadSnapshotPResponse> observer =
+          SnapshotDownloader.forFollower(mStorage, String.valueOf(client.getAddress()));
+      client.downloadSnapshot(observer);
+      return observer.getFuture().thenApplyAsync((termIndex) -> {
+        mDownloadedSnapshot = observer.getSnapshotToInstall();
+        transitionState(DownloadState.STREAM_DATA, DownloadState.DOWNLOADED);
```

Comment:
If you are referring to the state transition itself not being successful, it is not possible because the previous transition  `transitionState(DownloadState.IDLE, DownloadState.STREAM_DATA)` ensured the state is STREAM_DATA before this method is called. If we break this in the future, the transition will log a warning of the broken assumption.

If you are referring to the download being not successful. We have a `whenComplete` handler below to check the error thrown during download, log the error and reset the state.